"Dumb" auto import assist

eulerdisk opened the pull request:
This adds a new assist to "add xxx::yyy to the current file" when the cursor is on a PATH. It manages correctly nested imports,`self` keyword and creates new nested imports if necessary. [See the tests]
It doesn't use name resolution so in that sense is 'dumb', but I have plans to do that. That in the future will be useful to auto import trait names in autocompletion for example.

It can easily be extended to provide multiple actions to select in which scope to import. That's another thing I plan to do.

@matklad I copied some indentation code from `ide_light`, I don't know at the moment if/how you want to refactor that code. This assist was meant to be in `ide_light`.

matklad commented:
Haven't looked at the code yet, but the idea is 100% brilliant!

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+    SyntaxKind::{ PATH, PATH_SEGMENT, COLONCOLON, COMMA }
+};
+use crate::assist_ctx::{AssistCtx, Assist, AssistBuilder};
+use itertools::{ Itertools, EitherOrBoth };
```
An interesting next step along the lines of this PR is "optimize imports" action, which reformats imports, sorts `crate::` paths last and merges prefixes. 

Though perhaps to do that we'll need a "structured editing API", that is, something like "remove this import" rather than "replace this sub string with that sub string". There's no such API in ra_syntax yet, it would be fun to design and implement it.

eulerdisk replied:
Maybe having something like a mutable AST should be enough?
```
let mut mtreelist = treelist.get_mutable();
mtreelist.add_path(MutPath.from_str("foo::bar"))?
indent(mtreelist);
let diffs = diff(trellis, mstreelist); // get text insertions/deletions
```
Apart from `ident` and `diff`, I thing everything else can be autogenerated.

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+use itertools::{ Itertools, EitherOrBoth };
+
+// TODO: refactor this before merge
+mod formatting {
```
Yeah, formatting module in ra_ide_api_light is kinda random at the moment. Really, this should be a separate crate which is as powerful as rustfmt. 

You can use `formatting` from `ra_ide_api_light`, but I would also love a separate PR which creates a new `ra_fmt` crate and moves this module there.  I think we then would be able to drop dependency from assists on ide_api_light.

eulerdisk replied:
I'll go for another PR then, introducing `ra_fmt`.

matklad replied:
Awesome! Feel free to add/remove stuff from fmt module, it's currently a somewhat ad-hoc collections of functions, I imagine it should grow into a better "all formatting" API in the future!

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+    return Some(v);
+}
+
+fn collect_path_segments_raw<'b, 'a: 'b>(
```
I think it's OK to only declare `'a` lifetime and left `'b` to elision.

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+
+fn compare_path_segment(a: &ast::PathSegment, b: &ast::PathSegment) -> bool {
+    if let (Some(ka), Some(kb)) = (a.kind(), b.kind()) {
+        return match (ka, kb) {
```
Don't need this return. That's actually also a warning we can already implement.

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+}
+
+#[derive(Copy, Clone)]
+enum PathSegmentsMatch {
```
Hm, I wonder if instead of `enum` we could return a single `usize` -- length of the common prefix. It sort-of has the same amount of info as this `enum`, and doesn't suffer the problem that `Empy` and `Partial(0)` are the same.

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+    // Add a brand new use statement.
+    AddNewUse(
+        Option<&'a SyntaxNode>, // anchor node
+        bool,                   // true if we want to add the new statement after the anchor
```
Perhaps it makes sense to make this a struct variant, if we need a comment anyway? Positional bools are awkward .

matklad commented on crates/ra_assists/src/auto_import.rs:
```diff
@@ -0,0 +1,780 @@
+
+// Find out the best ImportAction to import target path against current_use_tree.
+// If current_use_tree has a nested import the function gets called recursively on every UseTree inside a UseTreeList.
+fn walk_use_tree_for_best_action<'b, 'c, 'a: 'b + 'c>(
```
ditto, 'b and 'c are not required I think.

matklad commented:
Haven't reviewed the bulk of the logic yet, but the general structure and tests look good to me!

One alternative approach here would be to always add a simple "fully-qualified" path first, and then invoking a separate "optimize imports" action to rewrite & merge all the imports.

eulerdisk commented:
Currently this assist is designed to modify existing code as little as possible.
I never thought  around "optimize imports" deeply, but I think it's a bit more complicated. It seems to me that it involves a lot of heuristics, user preferences and indentation capabilities, so I think it's a job for `rustfmt` (Maybe they are already doing it?).

[The best solution in the long run in having `rustfmt` implemented on top of `ra_syntax` anyway]

matklad commented:
>It seems to me that it involves a lot of heuristics, user preferences and indentation capabilities, so I think it's a job for `rustfmt` (Maybe they are already doing it?).

Agree, but we totally need to rewrite `rustfmt` on top of `ra_syntax`, and "optimizing imports" in particular seems an interesting hard+isolated task for prototyping editing API. We musn't do it in this PR of course, just something to keep in mind for future work.

eulerdisk commented:
Another thing "optimize imports" should do is to remove unused imports, so it's not entirely a `rustfmt` job.

eulerdisk commented:
@matklad I addressed your comments. I removed `PathSegmentsMatch` entirely, I'm not sure if it's better for you now. The `match` in `walk_use...` gets messier, so tell me if you want me to revert that.

matklad commented:
bors r+

Awesome!

matklad commented:
Opened https://github.com/rust-analyzer/rust-analyzer/issues/779 about well-typed editing API.

matklad commented:
Hey, @eulerdisk just want to let you know that this code have effectively grown to a full-blown auto import recently, thanks for writing it more than a year ago!
